README.md
Overview
===
Library for interacting with the SmugMug 1.3.0 API using OAuth authentication. This does not do any OAuth authorization or setup, it's assumed you're using another gem such as omniauth-oauth that handles that part.

Compability
-
Tested against Ruby 1.8.7, 1.9.2, 2.0.0 and JRuby, build history is available [here](http://travis-ci.org/zanker/ruby-smugmug).

<img src="https://secure.travis-ci.org/zanker/ruby-smugmug.png?branch=master&.png"/>

Examples
-
This is just a thin wrapper around the [SmugMug 1.3.0 API](http://wiki.smugmug.net/display/API/API+1.3.0), it's a 1:1 wrapper, so all of the documentation on the SmugMug page applies to this library. You can use any arguments that the SmugMug 1.3.0 documentation shows under the OAuth option.

    client = SmugMug::Client.new(:api_key => "1234-api", :oauth_secret => "4321-secret", :user => {:token => "abcd-token", :secret => "abcd-secret"})

    data = client.users.getStats(:Month => 2, :Year => 2012)
    puts data # {"Bytes"=>0, "Hits"=>0, "Large"=>0, "Medium"=>0, "Small"=>0, "Video110"=>0, "Video200"=>0, "Video320"=>0, "Video640"=>0, "X2Large"=>0, "X3Large"=>0, "XLarge"=>0}

    data = client.styles.getTemplates
    puts data # [{"id"=>0, "Name"=>"Viewer Controlled"}, {"id"=>3, "Name"=>"SmugMug"}, {"id"=>4, "Name"=>"Traditional"}, {"id"=>7, "Name"=>"All Thumbs"}, {"id"=>8, "Name"=>"Slideshow"}, {"id"=>9, "Name"=>"Journal (Old)"}, {"id"=>10, "Name"=>"SmugMug Small"}, {"id"=>11, "Name"=>"Filmstrip"}, {"id"=>12, "Name"=>"Critique"}, {"id"=>16, "Name"=>"Journal"}, {"id"=>17, "Name"=>"Thumbnails"}]


Because uploading is a special case and not under the same group of APIs, it's called slightly differently. See http://wiki.smugmug.net/display/API/Uploading for more information or the documentation linked below for a list of arguments.

    client = SmugMug::Client.new(:api_key => "1234-api", :oauth_secret => "4321-secret", :user => {:token => "abcd-token", :secret => "abcd-secret"})
    data = client.upload_media(:file => File.new("/Users/foobar/Desktop/image.jpeg", :AlbumID => 51343)
    puts data # {"id"=>1970029991, "Key"=>"rnSfAak", "URL"=>"http://foobar.smugmug.com/Other/Foo/51343_k8W1aR#1970029991_rnSfAak"}


Documentation
-
See http://rubydoc.info/github/zanker/ruby-smugmug/master/frames for full documentation.

License
-
Available under the MIT license, see LICENSE for more information.

CHANGELOG.md
# Overview

## 0.0.1 [In progress]




spec/spec_helper.rb
path = File.expand_path("../../", __FILE__)
require "#{path}/lib/ruby-smugmug.rb"

Dir["#{path}/spec/support/*.rb"].each {|file| require file}

RSpec.configure do |c|
  c.mock_with :rspec
end


spec/smugmug/http_spec.rb
require "spec_helper"

describe SmugMug::HTTP do
  include Support::ResponseMock

  it "uploads a file" do
    res_mock = mock("Response")
    res_mock.stub(:body).and_return(SmugMugResponses::FILE_SUCCESS)
    res_mock.stub(:code).and_return("200")
    res_mock.stub(:message).and_return("OK")
    res_mock.stub(:header).and_return({})

    http_mock = mock("HTTP")
    http_mock.should_receive(:request_post).with("/", "foo bar", hash_including("Content-Length" => "7", "Content-MD5" => "327b6f07435811239bc47e1544353273", "X-Smug-FileName" => "image.jpg", "X-Smug-AlbumID" => "1234", "Authorization" => "OAuth oauth_consumer_key=\"consumer\", oauth_nonce=\"nonce\", oauth_signature_method=\"sig-method\", oauth_signature=\"sig\", oauth_timestamp=\"1234\", oauth_version=\"1.0\", oauth_token=\"token\"")).and_return(res_mock)

    Net::HTTP.should_receive(:new).and_return(http_mock)

    http = SmugMug::HTTP.new(:api_key => "1234-api", :oauth_secret => "4321-secret", :user => {:token => "abcd-token", :secret => "abcd-secret"})
    http.should_receive(:sign_request).with("POST", SmugMug::HTTP::UPLOAD_URI, nil).and_return({"oauth_consumer_key" => "consumer", "oauth_nonce" => "nonce", "oauth_signature_method" => "sig-method", "oauth_signature" => "sig", "oauth_timestamp" => "1234", "oauth_version" => "1.0", "oauth_token" => "token"})

    data = http.request(:uploading, {:FileName => "image.jpg", :content => "foo bar", :AlbumID => 1234})
    data.should == JSON.parse(SmugMugResponses::FILE_SUCCESS)["Image"]
  end

  it "unzips gzipped responses" do
    mock_response(SmugMugResponses::SUCCESS)

    http = SmugMug::HTTP.new(:api_key => "1234-api", :oauth_secret => "4321-secret", :user => {:token => "abcd-token", :secret => "abcd-secret"})

    data = http.request("users.getInfo", {})
    data.should == JSON.parse(SmugMugResponses::SUCCESS)["User"]
  end

  it "unzips gzipped responses" do
    output = StringIO.new
    gz = Zlib::GzipWriter.new(output)
    gz.write(SmugMugResponses::SUCCESS)
    gz.close

    res_mock = mock("Response")
    res_mock.stub(:body).and_return(output.string)
    res_mock.stub(:code).and_return("200")
    res_mock.stub(:message).and_return("OK")
    res_mock.stub(:header).and_return({"content-encoding" => "gzip"})

    http_mock = mock("HTTP")
    http_mock.should_receive(:use_ssl=).with(true)
    http_mock.should_receive(:verify_mode=).with(anything)
    http_mock.should_receive(:request_post).with(anything, anything, anything).and_return(res_mock)

    Net::HTTP.should_receive(:new).and_return(http_mock)

    http = SmugMug::HTTP.new(:api_key => "1234-api", :oauth_secret => "4321-secret", :user => {:token => "abcd-token", :secret => "abcd-secret"})

    data = http.request("users.getInfo", {})
    data.should == JSON.parse(SmugMugResponses::SUCCESS)["User"]
  end

  it "handles response errors" do
    http = SmugMug::HTTP.new(:api_key => "1234-api", :oauth_secret => "4321-secret", :user => {:token => "abcd-token", :secret => "abcd-secret"})

    mock_response(SmugMugResponses::ERROR % 99)
    lambda { http.request("users.getInfo", {}) }.should raise_error(SmugMug::ReadonlyModeError)

    mock_response(SmugMugResponses::ERROR % 30)
    lambda { http.request("users.getInfo", {}) }.should raise_error(SmugMug::OAuthError)

    mock_response(SmugMugResponses::ERROR % 1)
    lambda { http.request("users.getInfo", {}) }.should raise_error(SmugMug::RequestError)
  end

  it "handles HTTP errors" do
    res_mock = mock("Response")
    res_mock.stub(:code).and_return("404")
    res_mock.stub(:message).and_return("Not Found")
    res_mock.stub(:header).and_return({})

    http_mock = mock("HTTP")
    http_mock.should_receive(:use_ssl=).with(true)
    http_mock.should_receive(:verify_mode=).with(anything)
    http_mock.should_receive(:request_post).with(anything, anything, anything).and_return(res_mock)

    Net::HTTP.should_receive(:new).and_return(http_mock)

    http = SmugMug::HTTP.new(:api_key => "1234-api", :oauth_secret => "4321-secret", :user => {:token => "abcd-token", :secret => "abcd-secret"})
    lambda { http.request("users.getInfo", {}) }.should raise_error(SmugMug::HTTPError)
  end

  it "generates a correct HMAC-SHA1 digest" do
    Time.stub(:now).and_return(Time.at(1341594000))

    nonce = Digest::MD5.hexdigest("foobar")
    Digest::MD5.should_receive(:hexdigest).and_return(nonce)

    http = SmugMug::HTTP.new(:api_key => "1234-api", :oauth_secret => "4321-secret", :user => {:token => "abcd-token", :secret => "abcd-secret"})
    postdata = http.sign_request("POST", SmugMug::HTTP::API_URI, {"method" => "smugmug.foo.bar", "a" => "Foo & Bar", "b" => 5, "c" => "Foo\nBar"})
    postdata.should == "a=Foo+%26+Bar&b=5&c=Foo%0ABar&method=smugmug.foo.bar&oauth_consumer_key=1234-api&oauth_nonce=3858f62230ac3c915f300c664312c63f&oauth_signature_method=HMAC-SHA1&oauth_timestamp=1341594000&oauth_token=abcd-token&oauth_version=1.0&oauth_signature=aACF%2BsJgVSyhWkhT%2BHlIwboZPSw%3D"
  end
end

spec/smugmug/client_spec.rb
require "spec_helper"

describe SmugMug::Client do
  before :each do
    @client = SmugMug::Client.new(:api_key => "1234-api", :oauth_secret => "4321-secret", :user => {:token => "abcd-token", :secret => "abcd-secret"})
  end

  it "creates a SmugMug API wrapper" do
    wrapper = @client.accounts
    wrapper.should be_a_kind_of(SmugMug::ApiCategory)
    wrapper.instance_variable_get(:@category).should == "accounts"
    @client.instance_variable_get(:@accounts_wrapper).should == wrapper
  end

  it "reuses created wrappers" do
    @client.users.should == @client.users
  end

  it "errors on invalid method" do
    lambda { @client.foobar }.should raise_error(NoMethodError)
  end

  it "prepares a File file for upload" do
    file = mock("File")
    file.should_receive(:is_a?).with(String).and_return(false)
    file.should_receive(:is_a?).with(File).and_return(true)
    file.should_receive(:path).and_return("/Users/foobar/Desktop/image.jpg")
    file.should_receive(:read).and_return("foo bar")

    http = mock("HTTP")
    http.should_receive(:request).with(:uploading, {:content => "foo bar", :FileName => "image.jpg", :AlbumID => 1234})
    @client.instance_variable_set(:@http, http)

    @client.upload_media(:file => file, :AlbumID => 1234)
  end

  it "prepares a String file for upload" do
    File.should_receive(:read).with("/Users/foobar/Desktop/image.jpg").and_return("foo bar")

    http = mock("HTTP")
    http.should_receive(:request).with(:uploading, {:content => "foo bar", :FileName => "image.jpg", :AlbumID => 1234})
    @client.instance_variable_set(:@http, http)

    @client.upload_media(:file => "/Users/foobar/Desktop/image.jpg", :AlbumID => 1234)
  end
end

lib/smugmug/client.rb
module SmugMug
  class Client
    ##
    # Creates a new client instance that can be used to access the SMugMug API
    # @param [Hash] args
    # @option args [String] :api_key Your SmugMug API key
    # @option args [String] :oauth_secret Your SmugMug OAuth secret key
    # @option args [Hash] :user Configuration for the user you are requesting/updating data for
    #   * :token [String] OAuth token for the user
    #   * :secret [String] OAuth secret token for the user
    # @option args [String, Optional :user_agent Helps SmugMug identify API calls
    # @option args [Hash, Optional] :http Additional configuration for the HTTP requests
    #   * :verify_mode [Integer, Optional] How to verify the SSL certificate when connecting through HTTPS, either OpenSSL::SSL::VERIFY_PEER or OpenSSL::SSL::VERIFY_NONE, defaults to OpenSSL::SSL::VERIFY_NONE
    #   * :ca_file [String, Optional] Path to the CA certification file in PEM format
    #   * :ca_path [String, Optional] Path to the directory containing CA certifications in PEM format
    #
    # @raise [ArgumentError]
    def initialize(args)
      raise ArgumentError, "API Key required" unless args.has_key?(:api_key)
      raise ArgumentError, "API OAuth secret required" unless args.has_key?(:oauth_secret)
      raise ArgumentError, "Must specify the users OAuth datA" unless args[:user].is_a?(Hash)
      raise ArgumentError, "Users OAuth token required" unless args[:user][:token]
      raise ArgumentError, "Users OAuth secret token required" unless args[:user][:secret]

      @http = HTTP.new(args)
    end

    ##
    # Uploading media files to SmugMug, see http://wiki.smugmug.net/display/API/Uploading for more information
    # @param [Hash] args
    # @option args [File] :file File or stream that can have the content read to upload
    # @option args [String] :file Binary contents of the file to upload
    # @option args [String] :FileName What the file name is, only required when passing :file as a string
    # @option args [Integer] :AlbumID SmugMug Album ID to upload the media to
    # @option args [Integer, Optional] :ImageID Image ID to replace if reuploading media rather than adding new
    # @option args [String, Optional] :Caption The caption for the media
    # @option args [Boolean, Optional] :Hidden Whether the media should be visible
    # @option args [String, Optional] :Keywords Keywords to tag the media as
    # @option args [Integer, Optional] :Altitude Altitude the media was taken at
    # @option args [Float, Optional] :Latitude Latitude the media was taken at
    # @option args [Float, Optional] :Longitude Latitude the media was taken at
    #
    # @raise [SmugMug::OAuthError]
    # @raise [SmugMug::HTTPError]
    # @raise [SmugMug::RequestError]
    # @raise [SmugMug::ReadonlyModeError]
    # @raise [SmugMug::UnknownAPIError]
    def upload_media(args)
      raise ArgumentError, "File is required" unless args.has_key?(:file)
      raise ArgumentError, "AlbumID is required" unless args.has_key?(:AlbumID)

      if args[:file].is_a?(String)
        args[:FileName] ||= File.basename(args[:file])
        args[:content] = File.read(args[:file])
      elsif args[:file].is_a?(File)
        args[:FileName] ||= File.basename(args[:file].path)
        args[:content] = args[:file].read
      else
        raise ArgumentError, "File must be a String or File"
      end

      args.delete(:file)
      @http.request(:uploading, args)
    end

    ##
    # Direct mapping of SmugMug 1.3.0 API, either see http://wiki.smugmug.net/display/API/API+1.3.0 or the README for examples
    #
    # @raise [SmugMug::OAuthError]
    # @raise [SmugMug::HTTPError]
    # @raise [SmugMug::RequestError]
    # @raise [SmugMug::ReadonlyModeError]
    # @raise [SmugMug::UnknownAPIError]
    def method_missing(method, *args)
      api_cat = method.to_s
      return super unless SmugMug::API_METHODS[api_cat]

      if klass = self.instance_variable_get("@#{api_cat}_wrapper")
        klass
      else
        self.instance_variable_set("@#{api_cat}_wrapper", SmugMug::ApiCategory.new(@http, api_cat))
      end
    end
  end
end

lib/smugmug/exceptions.rb
module SmugMug
  ##
  # Generic module that provides access to the code and text separately of the exception
  module ReplyErrors
    attr_reader :reply_text, :reply_code

    def initialize(msg, reply_code=nil, reply_text=nil)
      super(msg)
      @reply_code, @reply_text = reply_code, reply_text
    end
  end

  ##
  # Errors specific to the OAuth request
  class OAuthError < StandardError
    include ReplyErrors
  end

  ##
  # HTTP errors
  class HTTPError < StandardError
    include ReplyErrors
  end

  ##
  # Problem with the request
  class RequestError < StandardError
    include ReplyErrors

  end

  ##
  # SmugMug is in read-only mode
  class ReadonlyModeError < StandardError
  end

  ##
  # Trying to call an API that doesn't exist
  class UnknownAPIError < StandardError
  end
end

lib/smugmug/api_methods.rb
module SmugMug
  API_METHODS = {
    "accounts" => {"browse" => true},
    "albums" => {"applyWatermark" => true, "browse" => true, "changeSettings" => true, "comments" => true, "create" => true, "delete" => true, "get" => true, "getInfo" => true, "getStats" => true, "removeWatermark" => true, "reSort" => true},
    "albumtemplates" => {"changeSettings" => true, "create" => true, "delete" => true, "get" => true}, "auth" => {"checkAccessToken" => true, "getAccessToken" => true, "getRequestToken" => true},
    "categories" => {"create" => true, "delete" => true, "get" => true, "rename" => true}, "communities" => {"get" => true}, "coupons" => {"create" => true, "get" => true, "getInfo" => true, "modify" => true, "restrictions" => true},
    "family" => {"add" => true, "get" => true, "remove" => true, "removeAll" => true}, "fans" => {"get" => true}, "featured" => {"albums" => true}, "friends" => {"add" => true, "get" => true, "remove" => true, "removeAll" => true},
    "images" => {"applyWatermark" => true, "changePosition" => true, "changeSettings" => true, "collect" => true, "comments" => true, "crop" => true, "delete" => true, "get" => true, "getEXIF" => true, "getInfo" => true, "getStats" => true, "getURLs" => true, "removeWatermark" => true, "rotate" => true, "uploadFromURL" => true, "zoomThumbnail" => true},
    "printmarks" => {"create" => true, "delete" => true, "get" => true, "getInfo" => true, "modify" => true},
    "service" => {"ping" => true},
    "sharegroups" => {"albums" => true, "browse" => true, "create" => true, "delete" => true, "get" => true, "getInfo" => true, "modify" => true},
    "styles" => {"getTemplates" => true},
    "subcategories" => {"create" => true, "delete" => true, "get" => true, "getAll" => true, "rename" => true},
    "themes" => {"get" => true},
    "users" => {"getInfo" => true, "getStats" => true, "getTree" => true},
    "watermarks" => {"changeSettings" => true, "create" => true, "delete" => true, "get" => true, "getInfo" => true}
  }
end

lib/smugmug/http.rb
require "cgi"
require "openssl"
require "base64"
require "net/http"
require "json"

module SmugMug
  class HTTP
    API_URI = URI("https://api.smugmug.com/services/api/json/1.3.0")
    UPLOAD_URI = URI("http://upload.smugmug.com/")
    UPLOAD_HEADERS = [:AlbumID, :Caption, :Altitude, :ImageID, :Keywords, :Latitude, :Longitude, :Hidden, :FileName]
    OAUTH_ERRORS = {30 => true, 32 => true, 33 => true, 35 => true, 36 => true, 37 => true, 38 => true, 98 => true}

    ##
    # Creates a new HTTP wrapper to handle the network portions of the API requests
    # @param [Hash] args Same as [SmugMug::HTTP]
    #
    def initialize(args)
      @config = args
      @digest = OpenSSL::Digest::Digest.new("SHA1")

      @headers = {"Accept-Encoding" => "gzip"}
      if args[:user_agent]
        @headers["User-Agent"] = "#{args.delete(:user_agent)} (ruby-smugmug v#{SmugMug::VERSION})"
      else
        @headers["User-Agent"] = "Ruby-SmugMug v#{SmugMug::VERSION}"
      end
    end

    def request(api, args)
      uri = api == :uploading ? UPLOAD_URI : API_URI
      args[:method] = "smugmug.#{api}" unless api == :uploading

      http = ::Net::HTTP.new(uri.host, uri.port)
      http.set_debug_output(@config[:debug_output]) if @config[:debug_output]

      # Configure HTTPS if needed
      if uri.scheme == "https"
        http.use_ssl = true

        if @config[:http] and @config[:http][:verify_mode]
          http.verify_mode = @config[:http][:verify_mode]
          http.ca_file = @config[:http][:ca_file]
          http.ca_path = @config[:http][:ca_path]
        else
          http.verify_mode = OpenSSL::SSL::VERIFY_NONE
        end
      end

      # Upload request, which requires special handling
      if api == :uploading
        postdata = args.delete(:content)
        headers = @headers.merge("Content-Length" => postdata.length.to_s, "Content-MD5" => Digest::MD5.hexdigest(postdata), "X-Smug-Version" => "1.3.0", "X-Smug-ResponseType" => "JSON")

        UPLOAD_HEADERS.each do |key|
          next unless args[key] and args[key] != ""
          headers["X-Smug-#{key}"] = args[key].to_s
        end

        oauth = self.sign_request("POST", uri, nil)
        headers["Authorization"] = "OAuth oauth_consumer_key=\"#{oauth["oauth_consumer_key"]}\", oauth_nonce=\"#{oauth["oauth_nonce"]}\", oauth_signature_method=\"#{oauth["oauth_signature_method"]}\", oauth_signature=\"#{oauth["oauth_signature"]}\", oauth_timestamp=\"#{oauth["oauth_timestamp"]}\", oauth_version=\"#{oauth["oauth_version"]}\", oauth_token=\"#{oauth["oauth_token"]}\""

      # Normal API method
      else
        postdata = self.sign_request("POST", uri, args)
        headers = @headers
      end

      response = http.request_post(uri.request_uri, postdata, headers)
      if response.code == "204"
        return nil
      elsif response.code != "200"
        raise SmugMug::HTTPError.new("HTTP #{response.code}, #{response.message}", response.code, response.message)
      end

      # Check for GZIP encoding
      if response.header["content-encoding"] == "gzip"
        begin
          body = Zlib::GzipReader.new(StringIO.new(response.body)).read
        rescue Zlib::GzipFile::Error
          raise
        end
      else
        body = response.body
      end

      return nil if body == ""

      data = JSON.parse(body)

      if data["stat"] == "fail"
        # Special casing for SmugMug being in Read only mode
        if data["code"] == 99
          raise SmugMug::ReadonlyModeError.new("SmugMug is currently in read only mode, try again later")
        end

        klass = OAUTH_ERRORS[data["code"]] ? SmugMug::OAuthError : SmugMug::RequestError
        raise klass.new("Error ##{data["code"]}, #{data["message"]}", data["code"], data["message"])
      end

      data.delete("stat")
      data.delete("method")

      # smugmug.albums.changeSettings at the least doesn't return any data
      return nil if data.length == 0

      # It seems all smugmug APIs only return one hash of data, so this should be fine and not cause issues
      data.each do |_, value|
        return value
      end
    end

    ##
    # Generates an OAuth signature and updates the args with the required fields
    # @param [String] method HTTP method that the request is sent as
    # @param [String] uri Full URL of the request
    # @param [Hash] form_args Args to be passed to the server
    def sign_request(method, uri, form_args)
      # Convert non-string keys to strings so the sort works
      args = {}
      if form_args
        form_args.each do |key, value|
          next unless value and value != ""

          key = key.to_s unless key.is_a?(String)
          args[key] = value
        end
      end

      # Add the necessary OAuth args
      args["oauth_version"] = "1.0"
      args["oauth_consumer_key"] = @config[:api_key]
      args["oauth_nonce"] = Digest::MD5.hexdigest("#{Time.now.to_f}#{rand(10 ** 30)}")
      args["oauth_signature_method"] = "HMAC-SHA1"
      args["oauth_timestamp"] = Time.now.utc.to_i
      args["oauth_token"] = @config[:user][:token]

      # Sort the params
      sorted_args = []
      args.sort.each do |key, value|
        sorted_args.push("#{key.to_s}=#{CGI::escape(value.to_s)}")
      end

      postdata = sorted_args.join("&")

      # Final string to hash
      sig_base = "#{method}&#{CGI::escape("#{uri.scheme}://#{uri.host}#{uri.path}")}&#{CGI::escape(postdata)}"

      signature = OpenSSL::HMAC.digest(@digest, "#{@config[:oauth_secret]}&#{@config[:user][:secret]}", sig_base)
      signature = CGI::escape(Base64.encode64(signature).chomp)

      if uri == API_URI
        "#{postdata}&oauth_signature=#{signature}"
      else
        args["oauth_signature"] = signature
        args
      end
    end
  end
end